azkaban-common/src/main/java/azkaban/project/JdbcProjectImpl.java

```diff
@@ -814,19 +762,23 @@ public Flow fetchFlow(final Project project, final String flowId) throws Project
   }
 
   @Override
-  public List<Flow> fetchAllProjectFlows(final Project project) throws ProjectManagerException {
-    final ProjectFlowsResultHandler handler = new ProjectFlowsResultHandler();
-    List<Flow> flows = null;
+  public Map<Project, List<Flow>> fetchAllFlowsForProjects(final List<Project> projects) throws ProjectManagerException {
+    final SQLTransaction<Map<Project, List<Flow>>> transaction = transOperator -> {
+      Map<Project, List<Flow>> projectToFlows = new HashMap();
+      for (Project p : projects) {
+        projectToFlows.put(p, transOperator
+            .query(ProjectFlowsResultHandler.SELECT_ALL_PROJECT_FLOWS, new ProjectFlowsResultHandler(), p.getId(),
+                p.getVersion()));
+      }
+      return projectToFlows;
+    };
+
     try {
-      flows = this.dbOperator
-          .query(ProjectFlowsResultHandler.SELECT_ALL_PROJECT_FLOWS, handler, project.getId(),
-              project.getVersion());
+      return this.dbOperator.transaction(transaction);
     } catch (final SQLException e) {
       throw new ProjectManagerException(
-          "Error fetching flows from project " + project.getName() + " version " + project
-              .getVersion(), e);
+          "Error fetching flows for " + projects.size() + " project(s).", e);
```

Comment:
This is a step back in terms of debugging. Did you try failing the query and found out what information you could get?

Reply:
If the query fails, the underlying SQL error will still bubble up and be printed. The fact that we are moving individual queries into a transaction naturally means we will loose some granularity in error catching - but that's just a tradeoff with the performance increase. In reality, any issues with the query will most likely be due to connection issues, malformed query, or another problem that is not specific to a single row for a project, so the loss of granularity doesn't matter much.